core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -1482,6 +1479,15 @@ class Partition(val topicPartition: TopicPartition,
     } else {
       debug(s"Enqueued ISR change to state $newLeaderAndIsr after transition to $proposedIsrState")
     }
+
+    // some delayed operations may be unblocked after HW changed
+    isrUpdateCompleteFuture.thenAccept(leaderHWIncremented => {
```

Comment:
I think there is still race condition. "thenAccept" can still happen in the current thread, and then the deadlock could still happen.